### ISSUE TITLE
Native driver for Handy Tech braille displays

### ISSUE THREAD

**Issue by bramd** [nvaccess/nvda#7458]
Currently, there are many braille display drivers that depend on an external DLL from the vendor. The Handy Tech driver is one of those. The DLL used for this driver is relatively large to distribute and this COM server has caused some crashes, performance problems and hard to debug issues in the past.

Rewriting the Handy Tech driver in native Python might solve some open issues and allows for better integration in the future with braille autodetection and other NVDA features. Also, we don't have to wait on a new version of the COM server to add support for new Handy Tech displays.

Work on a native driver has started in https://github.com/bramd/nvda/tree/handytech-native

**Comment by leonardder** [comment]
Great initiative. I've seen it working quite well already.

The following issues could be tackled in this:

* #5806: Basic Braille 40: No Panning Commands
* #5774: Basic Braille 40: Wrong label for braille display keys 
* #4775: Support for HandyTech Active Tactile Control. This support should and will probably be very basic though

This driver might depend on functionality to be enabled or disabled in the GUI (#7452), e.g. ATC, braille input, etc.

**Comment by bramd** [comment]
### Current status of the handytech-native branch

Implemented:
* Braille output for modern USB serial and HID models
* Handling of keys for modern USB serial and HID models including routing
* All known (from BRLTTY) hardware IDs for USB devices
* A tree of classes to handle model specific keys/functions/protocol variations
  * All known models (from BRLTTY) with correct ID used in the protocol and number of cells
  * Reasonable keymap for most models

To do:
* Implement older braille protocol for BrailleStar, BrailleWave and probably other models
* Implement Bookworm support. A very old and very tiny display with different protocol (low priority)
* Add Bluetooth names
* Add basic ATC support to emit gestures, might need a braille touch input handling framework in NVDA to function well (low priority)
* Implement braille input
* Implement more advanced functions, such as (low priority/next version)
  * Syncing the clock of newer displays which support this
  * Getting service status, firmware version and other display info
  * Implement a fallback for unrecognized displays and drive the mbased on cell count that can be reported by newer displays


Testing:
* Braille output and all keys on Modular Evolution work, I use it as my daily driver for the last days. The key mappings correspond with what was there in the COM server/old driver and expand on that by mapping the numeric keypad
* @leonardder reported that it works as well with the Active Braille, which is a HID display. All keys on that model should work as well
* I have an old Braille Star at the office that I can test later

**Comment by bramd** [comment]
@jcsteh Do you have any ideas on differentiating between models in the gestureMap? Is there functionality to use something like sub identifiers (e.g. brl(ht+ModularEvolution88):b1, where ModularEvolution88 is the model of Handy Tech display? Ideally, if that gesture isn't bound, it will fall back to the generic brl(handyTech):b1 gesture.

**Comment by zstanecic** [comment]
@bramd @leonardder the initiative and the driver is great, it even fixes some bugs in the notepad plus plus using NVDA, i.e routing to a particular text field works in scintilla controls normally. but, as i use an older braille wave model as a hid with the rs232 serial on my computer:, there are some issues, with the log attached: panning keys are not working, and some other keys are not working (see the log) i used the latest updated handytech-netive branch
[nvda.txt](https://github.com/nvaccess/nvda/files/1201152/nvda.txt)

**Comment by zstanecic** [comment]
@bramd so, let me answer your question one by one:
•Did it give you the expected braille output?
•Is it able to fill all cells of the display? Type a line of 40 characters in Notepad for example (BrailleWave has 40 cells from the top of my head)
yes, the outputa as expected, has 40 cells, even it’s stated in the log viewer.
•Could you enter input help (nvda+1) and press all keys on the display one by one? Are the keys bound to the commands you would expect? I think the 8 braille dots should work, the left/right keys will not and the space bar won't work either.
the only keys from the eight dots keys forks are as follows: b4 and b5, moving by line forward /backward, b2, toggling between review /focus modes as the braille review modes, and the enter key, dot8 is working. for reference b8 , 
the following keys are not working: b1, b3, b5, b7.
the display is connected through a prolific com3 usb to serial converter.
the left key and right key, and also space is not working,
when you try to press panning buttons in help mode it says blank. Panning buttons are these located left and right before and after cursor routings, if you understand me what i want to say. 
•Please suggest key mappings that are absent and would make sense for the Braille WaveFor the braille wave i suggest the following mapping:
note: sadly this display never provided braille input!.
b1, backspace, b2, can be the toggling of focus modes, b3, i suggest context presentation switching b4, leave it as is, b5 is unmapped even with the com server, b6 is unmapped, b7 is unmapped, and b8 is enter.
left key from the space is shift tab, and the right key after the space is tab.
as i said, even in jaws this display never supported brailleimput of grade 2 and eight dots braille.
And, this is the display without bluetoot, with the firmware version 2.10

**Comment by bramd** [comment]
@zstanecic Thanks. I'll consider your suggestions for the b keys. This keys are on almost every Handy Tech display, so I have to think if this makes sense for all of them. Also, b4/b5 should be prev/next line, even with the COM server. Once braille input has been implemented in this driver, this should also work for your Braille Wave.

I just updated the branch with the keys for the Braille Wave. Please let me know if the mappings are as expected.

**Comment by zstanecic** [comment]
@bramd, now all the keys are not working on braille wave in the imput help, and outside of the imput help: see log again!
[nvda.txt](https://github.com/nvaccess/nvda/files/1201320/nvda.txt)

**Comment by bramd** [comment]
@zstanecic Ah, I seey, should be fixed.

**Comment by zstanecic** [comment]
@bramd 
i don’t see any new fixes in the repo yet...

**Comment by zstanecic** [comment]
ok. another steps to reproduce for keys b3 and b6: 1. press ins +1 2. press b3 or b6 these are not read outloud with the description only.. description of b1 was not read too...
panning is working as expected, escape is working as expected and the enter key on the display is working as expected i.e they act like tab and shift tab. 
and, ah i forgot
enter in the keymap is also emulated as escape+ return keys.
I don’t know, compared with the old comm server driver, the navigation by headings in chrome is slover when the driver is running

**Comment by bramd** [comment]
@zstanecic Thanks for the feedback. It is correct that b3 and b6 do not speak in input help. This is because they are mapped to left/right panning as well. This is needed for the Basic Braille, since it has only 6 b-keys. Escape+return has been mapped to enter as well.

**Comment by bramd** [comment]
Braille input has now been implemented. It can be turned on/off by pressing space+dots 1, 2, 7, that is space+b1+b3+b4 in the Handy Tech key names. This is consistent with other Handy Tech drivers that have this feature.

**Comment by zstanecic** [comment]
@bramd theres’ one thing: in the comm driver wich the stable nvda uses, theres’ no braille input for the braille wave device. Now the imput works on my device. I can just say thank you. now, it will be the big time for improvements in the braille tables, as some have big errors, like the serbian one, because when you try to write something, you cannot do it normally, it doesn’t respect the number signs. it’s depending of the table.
so the NVDA drivers wich you done, works now perfectly on the braille wave, without any error. I don’t know, how it will work with the actilino and other devices like braillino and braille star. I don’t have braillino or active braille to test how this thing works with these displays. For me, it will be great, to merge it for the next release, because it’s the best driver.

**Comment by zstanecic** [comment]
@bramd as the handytech displays are popular  in Croatia, can you give me the driver as an NVDA add-on so that i can test the driver in my steble envirinment? or give me instructions how this driver will work as an separate driver, because i want to distribute the driver to the people, who have more details, so that i can give you more feedback regarding the driver. All the best, Zvonimir

**Comment by ruifontes** [comment]
I can test with a Basic Braille 20.

Rui Fontes


-----Mensagem Original----- 
De: zstanecic
Data: 5 de agosto de 2017 02:16
Para: nvaccess/nvda
Cc: Subscribed
Assunto: Re: [nvaccess/nvda] Native driver for Handy Tech braille displays 
(#7458)



@bramd theres’ one thing: in the comm driver wich the stable nvda uses, 
theres’ no braille input for the braille wave device. Now the imput works on 
my device. I can just say thank you. now, it will be the big time for 
improvements in the braille tables, as some have big errors, like the 
serbian one, because when you try to write something, you cannot do it 
normally, it doesn’t respect the number signs. it’s depending of the table.
so the NVDA drivers wich you done, works now perfectly on the braille wave, 
without any error. I don’t know, how it will work with the actilino and 
other devices like braillino and braille star. I don’t have braillino or 
active braille to test how this thing works with these displays. For me, it 
will be great, to merge it for the next release, because it’s the best 
driver.

**Comment by jcsteh** [comment]
[**@bramd**](https://github.com/bramd) commented on [5 Aug 2017, 02:24 GMT+10](https://github.com/nvaccess/nvda/issues/7458#issuecomment-320292251 "2017-08-04T16:24:21Z - Replied by Github Reply Comments"):
> @jcsteh Do you have any ideas on differentiating between models in the gestureMap? Is there functionality to use something like sub identifiers (e.g. brl(ht+ModularEvolution88):b1, where ModularEvolution88 is the model of Handy Tech display?

There's no specific functionality for this at present, but it sounds reasonable. I'd probably suggest br(handyTech.ModularEvolution88):b1; i.e. keep the handyTech and use a full stop (or maybe a comma) to separate, since we tend to use plus to mean indeterminate order where gestures are concerned.

> Ideally, if that gesture isn't bound, it will fall back to the generic brl(handyTech):b1 gesture.

To do this, you'd need your gesture's identifiers property to return two identifiers: `["br(handyTech.ModularEvolution88):b1", "br(handyTech):b1"]`. The base BrailleDisplayGesture doesn't have any support for this, so you'd either have to enhance it thus or implement `identifiers` directly yourself. I'd recommend the former, since otherwise, you'll also have to implement support for braille input identifiers.

BrailleDisplayGesture currently has a source attribute which is used to get the part of the gesture source in parentheses. I guess you could add a model attribute which, if not None, would cause an extra identifier to be produced as above. That would mean all you have to do is set the model attribute. I think that should be pretty easy to implement.

> However, a major issue is how these new model specific gestures show up in the gui. The input gestures dialog uses inputCore.getDisplayTextForGestureIdentifier to get a proper readable name for the gesture.

Note that no display drivers currently return display text other than the internal names of the keys. So, I don't think supporting this is necessarily a requirement. If you did want to support it, though, you could also register the same gesture class for multiple gesture sources, since getDisplayTextForIdentifier also gives you the gesture source.

**Comment by bramd** [comment]
@ruifontes Are you able to run this from source? Keys are not mapped properly for the Basic braille yet, but it should work. 

@zstanecic I understand that testing a build or an add-on is easier. Unfortunately, it is a bit hard to package this as an add-on for now. First of all, there is already a driver named handyTech and this might conflict with gestures and reverting to the original driver will cause an error. This is due to a new port variable in the configuration for this driver. Also, we now depend on model specific key bindings/gestures, something that isn't in NVDA's release or master branch yet.

I'll try to arrange a try-build when all basic functionality for all models is done to ease testing.

**Comment by ruifontes** [comment]
Hi Bram!
Never did that, but if it not requires lots of Gb and you gave me the basics 
to set my machine to do that, I will be able to test it...

Rui


-----Mensagem Original----- 
De: Bram Duvigneau
Data: 12 de agosto de 2017 14:58
Para: nvaccess/nvda
Cc: ruifontes ; Mention
Assunto: Re: [nvaccess/nvda] Native driver for Handy Tech braille displays 
(#7458)



@ruifontes Are you able to run this from source? Keys are not mapped 
properly for the Basic braille yet, but it should work.

@zstanecic I understand that testing a build or an add-on is easier. 
Unfortunately, it is a bit hard to package this as an add-on for now. First 
of all, there is already a driver named handyTech and this might conflict 
with gestures and reverting to the original driver will cause an error. This 
is due to a new port variable in the configuration for this driver. Also, we 
now depend on model specific key bindings/gestures, something that isn't in 
NVDA's release or master branch yet.

I'll try to arrange a try-build when all basic functionality for all models 
is done to ease testing.

**Comment by bramd** [comment]
@zstanecic Could you please test the latest changes with your BrailleWave? According to BRLTTY code some displays are driven by an older protocol (BrailleWave, EasyBraille etc). Sending braille using the newer protocol worked, but at least the Easy Braille gave error beeps sometimes. Now the older protocol is used for this devices.

@ruifontes @leonardder indicated that he can trigger try builds on the NVAccess build system. We will do one of those very soon, so you can just test by running an executable like all other NVDA releases. If you want to run from source anyway, check the readme in the root of the repository for instructions.

**Comment by zstanecic** [comment]
@bramd now pulling changes

**Comment by zstanecic** [comment]
Hi, now the imput is not working: outpus at the beginning of each message returns k dots 1 3 
after pressing the imput key it breaks, and one fragment from the log, wich is important:
IO - braille.BrailleBuffer.update (20:44:26.519):
Braille regions text: [u'NVDA uruchomiony']
IO - braille.BrailleHandler.update (20:44:26.519):
Braille window dots: 13457 12367 1457 17 - 136 1235 136 14 125 135 134 24 135 1345 13456
DEBUG - core.main (20:44:26.523):
Initializing core pump
DEBUG - core.main (20:44:26.523):
Initializing watchdog
DEBUG - core.main (20:44:26.526):
Update checking not supported
INFO - core.main (20:44:26.526):
NVDA initialized
DEBUG - core.main (20:44:26.526):
entering wx application main loop
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:26.556):
ACK received
IO - speech.speak (20:44:26.726):
Speaking [LangChangeCommand ('hr'), u'TeamTalk 5 Classic v. 5.2.3  dialog  Statyczny tekst']
IO - speech.speak (20:44:26.760):
Speaking [LangChangeCommand ('hr'), u'Drzewo']
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:26.763):
Unhandled packet of type '\x05'
IO - speech.speak (20:44:26.775):
Speaking [LangChangeCommand ('hr'), u'poziom 0  Slavoland (1)  rozwini\u0119te  1 z 1']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:26.776):
Potential unimplemented child class: <NVDAObjects.Dynamic_DialogIAccessibleWindowNVDAObject object at 0x057AF790>
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:26.779):
Potential unimplemented child class: <NVDAObjects.IAccessible.sysTreeView32.TreeView object at 0x057AF7D0>
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:26.782):
Potential unimplemented child class: <NVDAObjects.Dynamic_TreeViewItemOutlineItemIAccessible object at 0x057AF850>
IO - braille.BrailleBuffer.update (20:44:26.783):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst ', u'wd ', u'Slavoland (1) - 1 z 1 lv 0']
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:29.421):
Unhandled packet of type '"'
IO - braille.BrailleHandler.update (20:44:30.513):
Braille window dots: 2345 15 13 234 2345 - 2456 145 - 2347 123 1 1236 135 123 1 1345 145 - 126 16 345 - 36 - 16 - 1356 - 16 - 123 1236 - 346
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:30.539):
Unhandled packet of type '!'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:30.954):
Unhandled packet of type '\xa1'
IO - inputCore.InputManager.executeGesture (20:44:36.512):
Input: kb(desktop):tab
IO - speech.speak (20:44:36.598):
Speaking [LangChangeCommand ('hr'), u'G\u0142o\u015bno\u015b\u0107 g\u0142\xf3wna  suwak  40']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:36.599):
Potential unimplemented child class: <NVDAObjects.IAccessible.IAccessible object at 0x057DFA50>
IO - braille.BrailleBuffer.update (20:44:36.601):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst ', u'G\u0142o\u015bno\u015b\u0107 g\u0142\xf3wna 40 suwak']
IO - braille.BrailleHandler.update (20:44:36.601):
Braille window dots: 12457 2378 135 357 1345 135 357 1468 - 1245 2378 267 2456 1345 1 - 1456 346 - 234 136 2456 1 13
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:36.638):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:36.842):
Unhandled packet of type '\x05'
IO - inputCore.InputManager.executeGesture (20:44:37.431):
Input: kb(desktop):shift+tab
IO - speech.speak (20:44:37.499):
Speaking [LangChangeCommand ('hr'), u'Drzewo']
IO - speech.speak (20:44:37.506):
Speaking [LangChangeCommand ('hr'), u'Slavoland (1)  rozwini\u0119te  1 z 1  poziom 0']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:37.506):
Potential unimplemented child class: <NVDAObjects.IAccessible.sysTreeView32.TreeView object at 0x057AF7F0>
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:37.509):
Potential unimplemented child class: <NVDAObjects.Dynamic_TreeViewItemOutlineItemIAccessible object at 0x057D4AD0>
IO - braille.BrailleBuffer.update (20:44:37.509):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst ', u'wd ', u'Slavoland (1) - 1 z 1 lv 0']
IO - braille.BrailleHandler.update (20:44:37.509):
Braille window dots: 2456 145 - 2347 123 1 1236 135 123 1 1345 145 - 126 16 345 - 36 - 16 - 1356 - 16 - 123 1236 - 346
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:37.546):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:37.747):
Unhandled packet of type '\x05'
IO - inputCore.InputManager.executeGesture (20:44:39.322):
Input: kb(desktop):tab
IO - speech.speak (20:44:39.359):
Speaking [LangChangeCommand ('hr'), u'G\u0142o\u015bno\u015b\u0107 g\u0142\xf3wna  suwak  40']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:39.361):
Potential unimplemented child class: <NVDAObjects.IAccessible.IAccessible object at 0x0561E370>
IO - braille.BrailleBuffer.update (20:44:39.362):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst ', u'G\u0142o\u015bno\u015b\u0107 g\u0142\xf3wna 40 suwak']
IO - braille.BrailleHandler.update (20:44:39.362):
Braille window dots: 12457 2378 135 357 1345 135 357 1468 - 1245 2378 267 2456 1345 1 - 1456 346 - 234 136 2456 1 13
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:39.400):
ACK received
IO - inputCore.InputManager.executeGesture (20:44:39.520):
Input: kb(desktop):tab
IO - speech.speak (20:44:39.599):
Speaking [LangChangeCommand ('hr'), u'G\u0142o\u015bno\u015b\u0107 mikrofonu  suwak  50']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:39.601):
Potential unimplemented child class: <NVDAObjects.IAccessible.IAccessible object at 0x057AF770>
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:39.601):
Unhandled packet of type '\x05'
IO - braille.BrailleBuffer.update (20:44:39.605):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst ', u'G\u0142o\u015bno\u015b\u0107 mikrofonu 50 suwak']
IO - braille.BrailleHandler.update (20:44:39.605):
Braille window dots: 12457 2378 135 357 1345 135 357 1468 - 134 24 13 1235 135 124 135 1345 136 - 156 346 - 234 136 2456 1 13
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:39.644):
ACK received
IO - inputCore.InputManager.executeGesture (20:44:39.709):
Input: kb(desktop):tab
IO - speech.speak (20:44:39.746):
Speaking [LangChangeCommand ('hr'), u'Pr\xf3g zadzia\u0142ania aktywacji g\u0142osem  suwak  0']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:39.746):
Potential unimplemented child class: <NVDAObjects.IAccessible.IAccessible object at 0x0561E490>
IO - braille.BrailleBuffer.update (20:44:39.747):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst ', u'Pr\xf3g zadzia\u0142ania aktywacji g\u0142osem 0 suwak']
IO - braille.BrailleHandler.update (20:44:39.749):
Braille window dots: 12347 1235 267 1245 - 1356 1 145 1356 24 1 2378 1 1345 24 1 - 1 13 2345 13456 2456 1 14 245 24 - 1245 2378 135 234 15 134 - 346 -
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:39.845):
Unhandled packet of type '\x05'
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:39.881):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.085):
Unhandled packet of type '\x05'
IO - inputCore.InputManager.executeGesture (20:44:40.213):
Input: kb(desktop):tab
IO - speech.speak (20:44:40.273):
Speaking [LangChangeCommand ('hr'), u'dialog z zak\u0142adkami']
IO - speech.speak (20:44:40.278):
Speaking [LangChangeCommand ('hr'), u'Czat  zak\u0142adka  zaznaczony']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:40.279):
Potential unimplemented child class: <NVDAObjects.IAccessible.IAccessible object at 0x0561E6B0>
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:40.282):
Potential unimplemented child class: <NVDAObjects.IAccessible.IAccessible object at 0x057FDD30>
IO - braille.BrailleBuffer.update (20:44:40.282):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst ', u'zak\u0142 ', u'Czat wbr zak\u0142adka']
IO - braille.BrailleHandler.update (20:44:40.282):
Braille window dots: 1356 1 13 2378 - 147 1356 1 2345 - 2456 12 1235 - 1356 1 13 2378 1 145 13 1
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.319):
ACK received
IO - inputCore.InputManager.executeGesture (20:44:40.513):
Input: kb(desktop):tab
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.523):
Unhandled packet of type '\x05'
IO - speech.speak (20:44:40.696):
Speaking [LangChangeCommand ('hr'), u'dialog']
IO - speech.speak (20:44:40.707):
Speaking [LangChangeCommand ('hr'), u'pole edycji  wielowierszowe  tylko do odczytu']
IO - speech.speak (20:44:40.743):
Speaking [LangChangeCommand ('hr'), u'zaznaczone 2334 znak\xf3w']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:40.746):
Potential unimplemented child class: <NVDAObjects.Dynamic_DialogIAccessibleWindowNVDAObject object at 0x057FD9D0>
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:40.749):
Potential unimplemented child class: <NVDAObjects.Dynamic_IAccessibleRichEdit20WindowNVDAObject object at 0x057FDE70>
IO - braille.BrailleBuffer.update (20:44:40.894):
Braille regions text: [u'* Do\u0142\u0105czono do kana\u0142u  ']
IO - braille.BrailleHandler.update (20:44:40.894):
Braille window dots: 3578 78 14578 13578 2378 5678 1478 135678 13578 134578 13578 78 14578 13578 78 1378 178 134578 178 2378 13678 78 78
IO - braille.BrailleHandler.update (20:44:40.895):
Braille window dots: 3578 78 14578 13578 2378 5678 1478 135678 13578 134578 13578 78 14578 13578 78 1378 178 134578 178 2378 13678 78 78
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.933):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.934):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.934):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.937):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.937):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.938):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.940):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.940):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.941):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.943):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:40.944):
Unhandled packet of type '\x05'
IO - inputCore.InputManager.executeGesture (20:44:41.088):
Input: kb(desktop):tab
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.144):
Unhandled packet of type '\x05'
IO - speech.speak (20:44:41.190):
Speaking [LangChangeCommand ('hr'), u'pole edycji']
IO - speech.speak (20:44:41.194):
Speaking [LangChangeCommand ('hr'), u'pusta']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:41.203):
Potential unimplemented child class: <NVDAObjects.Dynamic_IAccessibleEditWindowNVDAObject object at 0x05859850>
IO - braille.BrailleBuffer.update (20:44:41.213):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst ', u'zak\u0142 ', u'dlg ', u'edt ', ' ']
IO - braille.BrailleHandler.update (20:44:41.214):
Braille window dots: 15 145 2345 - -
IO - braille.BrailleHandler.update (20:44:41.217):
Braille window dots: 15 145 2345 - -
IO - braille.BrailleBuffer.update (20:44:41.224):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst ', u'zak\u0142 ', u'dlg ', u'edt ', ' ']
IO - braille.BrailleHandler.update (20:44:41.224):
Braille window dots: 15 145 2345 - -
IO - braille.BrailleHandler.update (20:44:41.226):
Braille window dots: 15 145 2345 - -
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.253):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.255):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.256):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.256):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.259):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.259):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.260):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.262):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.263):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.263):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.266):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.289):
Unhandled packet of type '\x05'
IO - inputCore.InputManager.executeGesture (20:44:41.684):
Input: kb(desktop):tab
IO - speech.speak (20:44:41.775):
Speaking [LangChangeCommand ('hr'), u'Drzewo']
IO - speech.speak (20:44:41.786):
Speaking [LangChangeCommand ('hr'), u'Slavoland (1)  rozwini\u0119te  1 z 1  poziom 0']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:41.789):
Potential unimplemented child class: <NVDAObjects.IAccessible.sysTreeView32.TreeView object at 0x05846430>
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:41.792):
Potential unimplemented child class: <NVDAObjects.Dynamic_TreeViewItemOutlineItemIAccessible object at 0x0585E050>
IO - braille.BrailleBuffer.update (20:44:41.792):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst ', u'wd ', u'Slavoland (1) - 1 z 1 lv 0']
IO - braille.BrailleHandler.update (20:44:41.795):
Braille window dots: 2456 145 - 2347 123 1 1236 135 123 1 1345 145 - 126 16 345 - 36 - 16 - 1356 - 16 - 123 1236 - 346
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:41.832):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:42.038):
Unhandled packet of type '\x05'
IO - inputCore.InputManager.executeGesture (20:44:59.519):
Input: kb(desktop):alt+tab
IO - speech.speak (20:44:59.822):
Speaking [LangChangeCommand ('hr'), u'rtmp://streamer54.interspace.mk:1935/mrt-radio/mr3 - VLC media player  linia 1  kolumna 2']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:44:59.825):
Potential unimplemented child class: <NVDAObjects.Dynamic_MultitaskingViewFrameListItemListItemUIA object at 0x05791A30>
IO - braille.BrailleBuffer.update (20:44:59.828):
Braille regions text: [u'rtmp://streamer54.interspace.mk:1935/mrt-radio/mr3 - VLC media player']
IO - braille.BrailleHandler.update (20:44:59.828):
Braille window dots: 1235 2345 134 1234 25 256 256 234 2345 1235 15 1 134 15 1235 156 1456 3 24 1345 2345 15 1235 234 1234 1 14 15 3 134 13 25 16 246 146 156 256 134 1235 2345
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:44:59.865):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:00.068):
Unhandled packet of type '\x05'
IO - inputCore.InputManager.executeGesture (20:45:00.098):
Input: kb(desktop):alt+tab
IO - speech.speak (20:45:00.148):
Speaking [LangChangeCommand ('hr'), u'TeamTalk 5 Classic v. 5.2.3  linia 1  kolumna 1']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:45:00.151):
Potential unimplemented child class: <NVDAObjects.Dynamic_MultitaskingViewFrameListItemListItemUIA object at 0x0585E050>
IO - braille.BrailleBuffer.update (20:45:00.153):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3']
IO - braille.BrailleHandler.update (20:45:00.154):
Braille window dots: 23457 15 1 134 23457 1 123 13 - 156 - 147 123 1 234 234 24 14 - 1236 3 - 156 3 34 3 146
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:00.190):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:00.394):
Unhandled packet of type '\x05'
IO - inputCore.InputManager.executeGesture (20:45:00.476):
Input: kb(desktop):alt+tab
IO - speech.speak (20:45:00.530):
Speaking [LangChangeCommand ('hr'), u'rtmp://streamer54.interspace.mk:1935/mrt-radio/mr3 - VLC media player  linia 1  kolumna 2']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:45:00.532):
Potential unimplemented child class: <NVDAObjects.Dynamic_MultitaskingViewFrameListItemListItemUIA object at 0x057DF8D0>
IO - braille.BrailleBuffer.update (20:45:00.533):
Braille regions text: [u'rtmp://streamer54.interspace.mk:1935/mrt-radio/mr3 - VLC media player']
IO - braille.BrailleHandler.update (20:45:00.533):
Braille window dots: 1235 2345 134 1234 25 256 256 234 2345 1235 15 1 134 15 1235 156 1456 3 24 1345 2345 15 1235 234 1234 1 14 15 3 134 13 25 16 246 146 156 256 134 1235 2345
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:00.569):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:00.773):
Unhandled packet of type '\x05'
IO - inputCore.InputManager.executeGesture (20:45:00.799):
Input: kb(desktop):alt+tab
IO - speech.speak (20:45:00.848):
Speaking [LangChangeCommand ('hr'), u'TeamTalk 5 Classic v. 5.2.3  linia 1  kolumna 1']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:45:00.851):
Potential unimplemented child class: <NVDAObjects.Dynamic_MultitaskingViewFrameListItemListItemUIA object at 0x0585E050>
IO - braille.BrailleBuffer.update (20:45:00.852):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3']
IO - braille.BrailleHandler.update (20:45:00.852):
Braille window dots: 23457 15 1 134 23457 1 123 13 - 156 - 147 123 1 234 234 24 14 - 1236 3 - 156 3 34 3 146
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:00.888):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:01.092):
Unhandled packet of type '\x05'
IO - speech.speak (20:45:01.191):
Speaking [LangChangeCommand ('hr'), u'TeamTalk 5 Classic v. 5.2.3  dialog  Statyczny tekst']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:45:01.194):
Potential unimplemented child class: <NVDAObjects.Dynamic_DialogIAccessibleWindowNVDAObject object at 0x058C5DB0>
IO - braille.BrailleBuffer.update (20:45:01.196):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst']
IO - braille.BrailleHandler.update (20:45:01.196):
Braille window dots: 23457 15 1 134 23457 1 123 13 - 156 - 147 123 1 234 234 24 14 - 1236 3 - 156 3 34 3 146 - 145 123 1245 -
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:01.232):
ACK received
IO - speech.speak (20:45:01.249):
Speaking [LangChangeCommand ('hr'), u'Drzewo']
IO - speech.speak (20:45:01.253):
Speaking [LangChangeCommand ('hr'), u'Slavoland (1)  rozwini\u0119te  1 z 1  poziom 0']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:45:01.339):
Potential unimplemented child class: <NVDAObjects.Dynamic_DialogIAccessibleWindowNVDAObject object at 0x058C5DB0>
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:45:01.344):
Potential unimplemented child class: <NVDAObjects.IAccessible.sysTreeView32.TreeView object at 0x057DF8D0>
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:45:01.346):
Potential unimplemented child class: <NVDAObjects.Dynamic_TreeViewItemOutlineItemIAccessible object at 0x05846430>
IO - braille.BrailleBuffer.update (20:45:01.348):
Braille regions text: [u'TeamTalk 5 Classic v. 5.2.3 dlg Statyczny tekst ', u'wd ', u'Slavoland (1) - 1 z 1 lv 0']
IO - braille.BrailleHandler.update (20:45:01.348):
Braille window dots: 2345 15 13 234 2345 - 2456 145 - 2347 123 1 1236 135 123 1 1345 145 - 126 16 345 - 36 - 16 - 1356 - 16 - 123 1236 - 346
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:01.436):
Unhandled packet of type '\x05'
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:01.471):
ACK received
IO - inputCore.InputManager.executeGesture (20:45:01.609):
Input: kb(desktop):windows+d
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:01.677):
Unhandled packet of type '\x05'
IO - speech.speak (20:45:01.892):
Speaking [LangChangeCommand ('hr'), u'FolderView  lista']
IO - speech.speak (20:45:01.895):
Speaking [LangChangeCommand ('hr'), u'Mozilla Thunderbird  8 z 25']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:45:01.898):
Potential unimplemented child class: <NVDAObjects.IAccessible.sysListView32.List object at 0x057FDE70>
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:45:01.900):
Potential unimplemented child class: <NVDAObjects.IAccessible.sysListView32.ListItem object at 0x058A2DD0>
IO - braille.BrailleBuffer.update (20:45:01.900):
Braille regions text: [u'FolderView lst ', u'Mozilla Thunderbird 8 z 25']
IO - braille.BrailleHandler.update (20:45:01.901):
Braille window dots: 123 234 2345 - 1347 135 1356 24 123 123 1 - 23457 125 136 1345 145 15 1235 12 24 1235 145 - 1256 - 1356 - 34 156
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:01.937):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:02.140):
Unhandled packet of type '\x05'
IO - inputCore.InputManager.executeGesture (20:45:02.237):
Input: kb(desktop):g
IO - speech.speak (20:45:03.467):
Speaking [LangChangeCommand ('hr'), u'Polski (Polska) - Polski + chorwacki']
IO - braille.BrailleBuffer.update (20:45:03.467):
Braille regions text: [u'Polski (Polska) - Polski + chorwacki']
IO - braille.BrailleHandler.update (20:45:03.467):
Braille window dots: 12347 135 123 234 13 24 - 126 12347 135 123 234 13 1 345 - 36 - 12347 135 123 234 13 24 - 235 - 14 125 135 1235 2456 1 14 13 24
IO - speech.speak (20:45:03.470):
Speaking [LangChangeCommand ('hr'), u'Polski + chorwacki']
IO - braille.BrailleBuffer.update (20:45:03.471):
Braille regions text: [u'Polski + chorwacki']
IO - braille.BrailleHandler.update (20:45:03.471):
Braille window dots: 12347 135 123 234 13 24 - 235 - 14 125 135 1235 2456 1 14 13 24
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.507):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.509):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.509):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.510):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.512):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.513):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.515):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.516):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.516):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.517):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.519):
Unhandled packet of type '\x05'
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:03.723):
Unhandled packet of type '\x05'
IO - inputCore.InputManager.executeGesture (20:45:04.638):
Input: kb(desktop):m
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:45:04.711):
Potential unimplemented child class: <NVDAObjects.IAccessible.sysListView32.List object at 0x057FDE70>
IO - braille.BrailleBuffer.update (20:45:04.719):
Braille regions text: [u'FolderView lst ', u'Mozilla Thunderbird 8 z 25']
IO - braille.BrailleHandler.update (20:45:04.719):
Braille window dots: 123 234 2345 - 1347 135 1356 24 123 123 1 - 23457 125 136 1345 145 15 1235 12 24 1235 145 - 1256 - 1356 - 34 156
IO - speech.speak (20:45:04.733):
Speaking [LangChangeCommand ('hr'), u'Miranda NG x64  22 z 25']
DEBUG - NVDAObjects.NVDAObject._get_placeholder (20:45:04.734):
Potential unimplemented child class: <NVDAObjects.IAccessible.sysListView32.ListItem object at 0x058C5FF0>
IO - braille.BrailleBuffer.update (20:45:04.736):
Braille regions text: [u'FolderView lst ', u'Miranda NG x64 22 z 25']
IO - braille.BrailleHandler.update (20:45:04.736):
Braille window dots: 1347 24 1235 1 1345 145 1 - 13457 12457 - 1346 1246 1456 - 34 34 - 1356 - 34 156
IO - braille.BrailleBuffer.update (20:45:04.742):
Braille regions text: [u'FolderView lst ', u'Miranda NG x64 22 z 25']
IO - braille.BrailleHandler.update (20:45:04.742):
Braille window dots: 1347 24 1235 1 1345 145 1 - 13457 12457 - 1346 1246 1456 - 34 34 - 1356 - 34 156
DEBUG - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:04.757):
ACK received
WARNING - brailleDisplayDrivers.handyTech.BrailleDisplayDriver._onReceive (20:45:04.759):
Unhandled packet of type '\x05'
and it’s crashing the device
i have continous beeps...

**Comment by bramd** [comment]
@zstanecic Thanks. Please check the latest changes for a fix to the braille output. I hope this also fixes the problems with input.

**Comment by zstanecic** [comment]
No, this is not fixing the imput, even i don’t have logged keys on the display. Now all the keys including routing doesn’t work, i know that, because they don’t interrupt speech.

**Comment by ruifontes** [comment]
Hi Bram, I will wait by the executable version...

I am engaged in 2017.3 translation...

Rui


-----Mensagem Original----- 
De: Bram Duvigneau
Data: 12 de agosto de 2017 19:23
Para: nvaccess/nvda
Cc: ruifontes ; Mention
Assunto: Re: [nvaccess/nvda] Native driver for Handy Tech braille displays 
(#7458)



@zstanecic Could you please test the latest changes with your BrailleWave? 
According to BRLTTY code some displays are driven by an older protocol 
(BrailleWave, EasyBraille etc). Sending braille using the newer protocol 
worked, but at least the Easy Braille gave error beeps sometimes. Now the 
older protocol is used for this devices.

@ruifontes @leonardder indicated that he can trigger try builds on the 
NVAccess build system. We will do one of those very soon, so you can just 
test by running an executable like all other NVDA releases. If you want to 
run from source anyway, check the readme in the root of the repository for 
instructions.

**Comment by bramd** [comment]
@zstanecic I think input is fixed again. Please post a log if not.

**Comment by leonardder** [comment]
@bramd I wonder whether we should do a separate pr for the model based gesture identifiers?

**Comment by bramd** [comment]
@leonardder Do you have any use for this outside the Handy Tech driver? A separate PR would be fine with me, but code that isn't used anywhere shouldn't be in core yet in my oppinion.

**Comment by leonardder** [comment]
[**@bramd**](https://github.com/bramd) commented on [22 aug. 2017 08:02 CEST](https://github.com/nvaccess/nvda/issues/7458#issuecomment-323927623 "2017-08-22T06:02:43Z - Replied by Github Reply Comments"):
> @leonardder Do you have any use for this outside the Handy Tech driver? A separate PR would be fine with me, but code that isn't used anywhere shouldn't be in core yet in my oppinion.

I'm sure other drivers can benefit from this, especially the Baum driver which supports a wide range of displays. See also https://github.com/nvaccess/nvda/issues/6063#issuecomment-319168473

**Comment by zstanecic** [comment]
Hi, can someone create the binary build of the NVDA version with this implementation?
as i have reinstalled the system and wasn’t active long time.
@leonardder @bramd

**Comment by bramd** [comment]
@zstanecic I thik the current version is ready for real world testing, so a binary build will come soon.

**Comment by bramd** [comment]
### Status

* After checking with Handy Tech I drop the older protocol for now. This will only affect the Bookworm (very old display), Braillino and Modular serie with old firmware. The Modular can be updated, the Braillino can't it seems. These displays can still be used through the Handy Tech COM server and their add-on or BRLTTY.
* The driver should work with all supported displays through USB, serial and Bluetooth. This needs field testing with various displays though.
* Braille input is fully implemented
* We enable ATC, so displays which support this can awake by touching the cells, which is done in hardware. We ignore ATC packets for now.
* There seems to be another USB chip in use on some old models(GoHubs). We need to test this and see if there is a way to include these in automatic port detection.

**Comment by zstanecic** [comment]
Hi!

I'm  waiting for the real world version

If there will be a new bug-free thing it will be great


W dniu 03.09.2017 o 14:53, Bram Duvigneau pisze:

**Comment by zstanecic** [comment]
Can the COM server be updated in the local package, i mean the dll?

i’m wondering how this will affect the stability or other things


W dniu 03.09.2017 o 15:04, Bram Duvigneau pisze:

**Comment by jcsteh** [comment]
The one thing I'd point out is that I'd be very reluctant to include both
the old and new drivers with NVDA itself, as it's just far too confusing
and creates an extra maintenance burden. I think we either need to drop the
old driver or distribute the new one as an add-on until we're reasonably
confident it can replace the old one.

**Comment by dcbahr** [comment]
I have a handy tech active braille, firmware v4.0. I would be willing to do some testing although I couldn't tell you what code would need to be changed as, well, I don't know code. But I know that i'd really like to be able to use the ATC on the active braille.

**Comment by bramd** [comment]
@dcbahr We are already testing with an Active Braille, but more testing before introducing this in master/release is always good. Please follow this issue to be notified about test builds and changes. ATC will not work in a first version, since it is not enough to just recognize the data from the device, NVDA needs to process that and determine if/when to scroll to the next line etc. However, I would like to add it later for sure.

**Comment by bramd** [comment]
I'm having issues with the Basic Braille displays. They lock up after writing lots of cells in quick succession. I suspect it is because writes take longer with this display, but I'm not yet sure about that. When I open Notepad and hold a key, for example "=", after 20 cells or so the display stops updating smoothly. If I hold it long enough, the driver/display crashes.

It seems the thread seems to spend to much time in writing and receiving new cells from NVDA, so it can't process the ACK responses. A proper solution for this would be to implement a write, read and wait for ACK, write next packet loop I think, so writing data and reading the ACK responses are balanced. Other protocols might also benefit from this, since we now just write data and ignore ACKs in the drivers, waiting for an ACK and then sending the next packet seems how these protocols are meant to be used.

@jcsteh Any ideas how I can debug this? I don't really get the interaction between our background thread, the events coming from Windows for writing/reading data and NVDA scheduling new cells on the thread to be displayed and it's not trivial to me how to implement a proper loop with a queue of packets to be send to the display with this asynchronous reading.

**Comment by ruifontes** [comment]
Do you need me to test with my BB 20?

Rui


-----Mensagem Original----- 
De: Bram Duvigneau
Data: 10 de setembro de 2017 15:08
Para: nvaccess/nvda
Cc: ruifontes ; Mention
Assunto: Re: [nvaccess/nvda] Native driver for Handy Tech braille displays 
(#7458)



I'm having issues with the Basic Braille displays. They lock up after 
writing lots of cells in quick succession. I suspect it is because writes 
take longer with this display, but I'm not yet sure about that. When I open 
Notepad and hold a key, for example "=", after 20 cells or so the display 
stops updating smoothly. If I hold it long enough, the driver/display 
crashes.

It seems the thread seems to spend to much time in writing and receiving new 
cells from NVDA, so it can't process the ACK responses. A proper solution 
for this would be to implement a write, read and wait for ACK, write next 
packet loop I think, so writing data and reading the ACK responses are 
balanced. Other protocols might also benefit from this, since we now just 
write data and ignore ACKs in the drivers, waiting for an ACK and then 
sending the next packet seems how these protocols are meant to be used.

@jcsteh Any ideas how I can debug this? I don't really get the interaction 
between our background thread, the events coming from Windows for 
writing/reading data and NVDA scheduling new cells on the thread to be 
displayed and it's not trivial to me how to implement a proper loop with a 
queue of packets to be send to the display with this asynchronous reading.

**Comment by dcbahr** [comment]
Well, I'm not sure where it's coming from, but I keep getting the message that my display needs servicing by my dealer. It lists the last service date as 2013, which is when I got the display. I have no clue how to turn this alert off, happens every time I load nvda. If you've got any ideas on this or if a beta of your driver will maybe fix this problem, that would be great. Derek is a friend of mine. 

But yes, getting ATC into a driver would be fantastic. But difficult, I'm guessing.

**Comment by jcsteh** [comment]
[**@bramd**](https://github.com/bramd) commented on [11 Sep 2017, 00:08 GMT+10](https://github.com/nvaccess/nvda/issues/7458#issuecomment-328344995 "2017-09-10T14:08:24Z - Replied by Github Reply Comments"):
> If I hold it long enough, the driver/display crashes.

When you say "crashes", does it just stop displaying stuff or do you get exceptions?

> It seems the thread seems to spend to much time in writing and receiving new cells from NVDA, so it can't process the ACK responses.

I guess that's possible. However:

1. Our write APCs shouldn't get any higher priority than the read (onReceive) callbacks.
2. We don't queue another write APC if we've already queued one and it hasn't executed yet. So, we don't pointlessly flood the APC queue with writes that will never do anything.

> A proper solution for this would be to implement a write, read and wait for ACK, write next packet loop I think, so writing data and reading the ACK responses are balanced. Other protocols might also benefit from this, since we now just write data and ignore ACKs in the drivers, waiting for an ACK and then sending the next packet seems how these protocols are meant to be used.

Interestingly, none of the protocols I've worked on so far (Baum, Brailliant, BrailleNote) have ACK responses. That seems rather strange and pointless to me, but if they do ACK every packet, that does suggest buffers could get full somewhere. Do you know of other protocols which ACK every packet?

> @jcsteh Any ideas how I can debug this? I don't really get the interaction between our background thread, the events coming from Windows for writing/reading data and NVDA scheduling new cells on the thread to be displayed

1. You can turn on [hwIo debug logging](https://github.com/nvaccess/nvda/wiki/DebuggingBrailleDisplayDrivers). That will log (at level debug) all data in and out, among other things. That should give you some idea of how many packets are going out versus how many packets are coming in.
2. If the crash you describe is actually a freeze (i.e. no output on the display and nothing coming in), it'd be good to know what function we get stuck in. You might need to add some additional debug logging at the start and end of `hwIo.IoBase._asyncRead`, `hwIo.IoBase._ioDone`, `hwIo.Serial.read` and `hwIo.Serial.write`.

> and it's not trivial to me how to implement a proper loop with a queue of packets to be send to the display with this asynchronous reading.

Unfortunately, the kind of loop you're talking about is kind of ugly async. Also, the whole point of doing background writes was to prevent slow displays from slowing down NVDA if there were lots of quick writes, since previously, we had to wait for every single write to finish. The idea is that writes that occurred while the display was finishing a previous write are useless except for the last write, since the user won't have a chance to even read the writes in the middle.

You could do something like this:

1. Have a flag for whether you are awaiting an ACK. Also, have a variable for pending cells to be written after the ACK. You only want to track the latest string of cells. Otherwise, you'll queue up a huge batch of writes and the display will be way behind the user.
2. When displaying cells, if you're awaiting an ACK, just set the pending cells variable. If you're not awaiting an ACK, display the cells and set the awaiting flag.
3. In your onReceive callback, if you receive an ACK, write pending cells (if any), clear the pending cells and clear the awaiting flag.

If other protocols require ACKs, we should consider abstracting this in core using a similar technique. That should be possible in `braille._BgThread`, but you'll also need a class variable specifying whether drivers need to wait for ACKs and also a notification (`extensionPoints.Action`) so displays can notify when they receive an ACK.

**Comment by leonardder** [comment]
I wonder whether we should change from GetOverlappedResult to GetOverlappedResultEx on Windows 8 and above. One advantage of this is that you can specify a timeout in GetOverlappedResultEx. Another advantage is that reading is not interrupted while waiting for a write to complete. See https://github.com/nvaccess/nvda/issues/1271#issuecomment-327700011

**Comment by jcsteh** [comment]
Oh. That's a potentially interesting solution to this problem: allowing reads while writes are in progress. However, the problem is that write APCs from the main thread will also get handled in the alertable wait, which is problematic; you'll end up with writes during writes unless you special case re-entrance somehow.

I'm also not thrilled about it being Windows >= 8 only. However, you should be able to achieve the same result on all Windows as follows:

1. Specify an event for write completion in `IoBase._writeOl`.
2. In `IoBase.write`, instead of `GetOverlappedResult`, use `WaitForSingleObjectEx` to wait on the write event with an alertable wait.

There's one further problem: `hwIo.Serial.write` doesn't use `IoBase.write`. Instead, it calls pyserial's implementation. You would need to look at what that does and work out whether it's safe to switch it to just use `IoBase.write`.

**Comment by leonardder** [comment]
[**@jcsteh**](https://github.com/jcsteh) commented on [11 sep. 2017 07:29 CEST](https://github.com/nvaccess/nvda/issues/7458#issuecomment-328419535 "2017-09-11T05:29:58Z - Replied by Github Reply Comments"):
> you should be able to achieve the same result on all Windows as follows:
> 
> 1. Specify an event for write completion in `IoBase._writeOl`.
> 2. In `IoBase.write`, instead of `GetOverlappedResult`, use `WaitForSingleObjectEx` to wait on the write event with an alertable wait.

Sounds to be a better solution indeed.

> There's one further problem: `hwIo.Serial.write` doesn't use `IoBase.write`. Instead, it calls pyserial's implementation. You would need to look at what that does and work out whether it's safe to switch it to just use `IoBase.write`.

Interesting enough, pyserial uses GetOverlappedResult, but one line above that, they commented out WaitForSingleObject. Anyway, I think it is quite safe to switch to IoBase.write, pyserial has some py3.1 workarounds as well as some additional error handling.

**Comment by leonardder** [comment]
[**@jcsteh**](https://github.com/jcsteh) commented on [11 sep. 2017 01:46 CEST](https://github.com/nvaccess/nvda/issues/7458#issuecomment-328382046 "2017-09-10T23:46:30Z - Replied by Github Reply Comments"):
> 1. Have a flag for whether you are awaiting an ACK. Also, have a variable for pending cells to be written after the ACK. You only want to track the latest string of cells. Otherwise, you'll queue up a huge batch of writes and the display will be way behind the user.
> 2. When displaying cells, if you're awaiting an ACK, just set the pending cells variable. If you're not awaiting an ACK, display the cells and set the awaiting flag.
> 3. In your onReceive callback, if you receive an ACK, write pending cells (if any), clear the pending cells and clear the awaiting flag.

I can report that this approach works as expected on a BB40.

**Comment by bramd** [comment]
[**@jcsteh**](https://github.com/jcsteh) commented on [Sep 11, 2017, 1:46 AM GMT+2](https://github.com/nvaccess/nvda/issues/7458#issuecomment-328382046 "2017-09-10T23:46:30Z - Replied by Github Reply Comments"):
> [**@bramd**](https://github.com/bramd) commented on [11 Sep 2017, 00:08 GMT+10](https://github.com/nvaccess/nvda/issues/7458#issuecomment-328344995):
> 
> > If I hold it long enough, the driver/display crashes.
> 
> When you say "crashes", does it just stop displaying stuff or do you get exceptions?
It didn't give a clear exception, but you would see the writes continue and less and less reads come in. Then NVDA disables the driver because displaying new cells fails. Why this happened was not clear.

> Interestingly, none of the protocols I've worked on so far (Baum, Brailliant, BrailleNote) have ACK responses. That seems rather strange and pointless to me, but if they do ACK every packet, that does suggest buffers could get full somewhere. Do you know of other protocols which ACK every packet?

I thought Baum did it from the top of my head, but just checked BRLTTY source and this does not seem to be the case. SO, let's just keep this local to the Handy Tech driver for now.

 
> You could do something like this:
> 
> 1.  Have a flag for whether you are awaiting an ACK. Also, have a variable for pending cells to be written after the ACK. You only want to track the latest string of cells. Otherwise, you'll queue up a huge batch of writes and the display will be way behind the user.
> 2.  When displaying cells, if you're awaiting an ACK, just set the pending cells variable. If you're not awaiting an ACK, display the cells and set the awaiting flag.
> 3.  In your onReceive callback, if you receive an ACK, write pending cells (if any), clear the pending cells and clear the awaiting flag.

I had an experiment doing almost that, but without the clearing of pendingCells and just a queue of packets to be written. Your suggested approach as implemented by @leonardder seems to work. I confirmed this on a Basic Braille 40 I have here for testing today.

**Comment by FelixGruetzmacher** [comment]
On one of my machines when starting NVDA with the native driver I get the following as part of the log:
INFO - external:brailleDisplayDrivers.handyTech.BrailleDisplayDriver.__init__ (17:25:32.986):
Found Active Braille connected via USB HID (\\?\hid#vid_1fe4&pid_0054&mi_00#8&38e8f3ff&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})
ERROR - stderr (17:25:32.986):
Exception
ERROR - stderr (17:25:33.005):
AttributeError
ERROR - stderr (17:25:33.023):
:
ERROR - stderr (17:25:33.039):
"'Hid' object has no attribute '_file'"
ERROR - stderr (17:25:33.055):
 in
ERROR - stderr (17:25:33.071):
<bound method Hid.__del__ of <hwIo.Hid object at 0x06D575B0>>
ERROR - stderr (17:25:33.086):
 ignored

Also on this machine, when going to Braille preferences the dialog fails to open but I get:
ERROR - braille.getDisplayList (17:51:47.184):
Traceback (most recent call last):
  File "braille.pyc", line 300, in getDisplayList
  File "brailleDisplayDrivers\brailliantB.pyc", line 108, in check
  File "brailleDisplayDrivers\brailliantB.pyc", line 91, in _getPorts
  File "hwPortUtils.pyc", line 236, in listComPorts
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 31: ordinal not in range(128)
ERROR - unhandled exception (17:51:47.529):
Traceback (most recent call last):
  File "gui\__init__.pyc", line 222, in onBrailleCommand
  File "gui\__init__.pyc", line 184, in _popupSettingsDialog
  File "gui\settingsDialogs.pyc", line 81, in __init__
  File "gui\settingsDialogs.pyc", line 1562, in makeSettings
  File "gui\settingsDialogs.pyc", line 1726, in updatePossiblePorts
  File "C:\Users\const\AppData\Roaming\nvda\brailleDisplayDrivers\handyTech.py", line 486, in getPossiblePorts
  File "hwPortUtils.pyc", line 236, in listComPorts
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 31: ordinal not in range(128)

Maybe something to do with localization and getting the friendly name. Crystal ball off. I'll do what I can to help.

**Comment by leonardder** [comment]
[**@FelixGruetzmacher**](https://github.com/FelixGruetzmacher) commented on [29 Sep 2017, 17:55 CEST](https://github.com/nvaccess/nvda/issues/7458#issuecomment-333165506 "2017-09-29T15:55:01Z - Replied by Github Reply Comments"):
> On one of my machines when starting NVDA with the native driver I get the following as part of the log:
> INFO - external:brailleDisplayDrivers.handyTech.BrailleDisplayDriver.__init__ (17:25:32.986):
> Found Active Braille connected via USB HID (\\?\hid#vid_1fe4&pid_0054&mi_00#8&38e8f3ff&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})
> ERROR - stderr (17:25:32.986):
> Exception
> ERROR - stderr (17:25:33.005):
> AttributeError
> ERROR - stderr (17:25:33.023):
> :
> ERROR - stderr (17:25:33.039):
> "'Hid' object has no attribute '_file'"
> ERROR - stderr (17:25:33.055):
>  in
> ERROR - stderr (17:25:33.071):
> <bound method Hid.__del__ of <hwIo.Hid object at 0x06D575B0>>
> ERROR - stderr (17:25:33.086):
>  ignored

Ah yes, that one. You can actually safely ignore that, its hwIo trying to close a port which is never opened. It is not driver specific and not at all live threatening.

Actually, the reason behind this is that the driver tries to connect to the keyboard HID device of the Active Braille, but the system doesn't allow access to that device anyway, so that's also fine.

> Also on this machine, when going to Braille preferences the dialog fails to open but I get:
> ERROR - braille.getDisplayList (17:51:47.184):
> Traceback (most recent call last):
>   File "braille.pyc", line 300, in getDisplayList
>   File "brailleDisplayDrivers\brailliantB.pyc", line 108, in check
>   File "brailleDisplayDrivers\brailliantB.pyc", line 91, in _getPorts
>   File "hwPortUtils.pyc", line 236, in listComPorts
> UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 31: ordinal not in range(128)
> ERROR - unhandled exception (17:51:47.529):
> Traceback (most recent call last):
>   File "gui\__init__.pyc", line 222, in onBrailleCommand
>   File "gui\__init__.pyc", line 184, in _popupSettingsDialog
>   File "gui\settingsDialogs.pyc", line 81, in __init__
>   File "gui\settingsDialogs.pyc", line 1562, in makeSettings
>   File "gui\settingsDialogs.pyc", line 1726, in updatePossiblePorts
>   File "C:\Users\const\AppData\Roaming\nvda\brailleDisplayDrivers\handyTech.py", line 486, in getPossiblePorts
>   File "hwPortUtils.pyc", line 236, in listComPorts
> UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 31: ordinal not in range(128)
> 
> Maybe something to do with localization and getting the friendly name. Crystal ball off. I'll do what I can to help.

This seems to be a line related to logging, I'll also see waht I can do.

**Comment by leonardder** [comment]
[**@FelixGruetzmacher**](https://github.com/FelixGruetzmacher) commented on [29 Sep 2017, 17:55 CEST](https://github.com/nvaccess/nvda/issues/7458#issuecomment-333165506 "2017-09-29T15:55:01Z - Replied by Github Reply Comments"):
> On one of my machines when starting NVDA with the native driver I get the following as part of the log:
> INFO - external:brailleDisplayDrivers.handyTech.BrailleDisplayDriver.__init__ (17:25:32.986):
> Found Active Braille connected via USB HID (\\?\hid#vid_1fe4&pid_0054&mi_00#8&38e8f3ff&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030})
> ERROR - stderr (17:25:32.986):
> Exception
> ERROR - stderr (17:25:33.005):
> AttributeError
> ERROR - stderr (17:25:33.023):
> :
> ERROR - stderr (17:25:33.039):
> "'Hid' object has no attribute '_file'"
> ERROR - stderr (17:25:33.055):
>  in
> ERROR - stderr (17:25:33.071):
> <bound method Hid.__del__ of <hwIo.Hid object at 0x06D575B0>>
> ERROR - stderr (17:25:33.086):
>  ignored

Ah yes, that one. You can actually safely ignore that, its hwIo trying to close a port which is never opened. It is not driver specific and not at all live threatening.

Actually, the reason behind this is that the driver tries to connect to the keyboard HID device of the Active Braille, but the system doesn't allow access to that device anyway, so that's also fine.

I just posted a commit that should deal with this more gracefully.

> Also on this machine, when going to Braille preferences the dialog fails to open but I get:
> ERROR - braille.getDisplayList (17:51:47.184):
> Traceback (most recent call last):
>   File "braille.pyc", line 300, in getDisplayList
>   File "brailleDisplayDrivers\brailliantB.pyc", line 108, in check
>   File "brailleDisplayDrivers\brailliantB.pyc", line 91, in _getPorts
>   File "hwPortUtils.pyc", line 236, in listComPorts
> UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 31: ordinal not in range(128)
> ERROR - unhandled exception (17:51:47.529):
> Traceback (most recent call last):
>   File "gui\__init__.pyc", line 222, in onBrailleCommand
>   File "gui\__init__.pyc", line 184, in _popupSettingsDialog
>   File "gui\settingsDialogs.pyc", line 81, in __init__
>   File "gui\settingsDialogs.pyc", line 1562, in makeSettings
>   File "gui\settingsDialogs.pyc", line 1726, in updatePossiblePorts
>   File "C:\Users\const\AppData\Roaming\nvda\brailleDisplayDrivers\handyTech.py", line 486, in getPossiblePorts
>   File "hwPortUtils.pyc", line 236, in listComPorts
> UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 31: ordinal not in range(128)
> 
> Maybe something to do with localization and getting the friendly name. Crystal ball off. I'll do what I can to help.

This seems to be a line related to logging.

**Comment by leonardder** [comment]
@FelixGruetzmacher: Have you a working source copy? If so, could you run this from source and paste the debug info? That will also log the line of failing code. Also, you could try to run with this particular statement commented out:

```
log.debugWarning("Couldn't get SPDRP_FRIENDLYNAME for %s: %s" % (port, ctypes.WinError()))
```

**Comment by dkager** [comment]
Sounds like an incarnation of #6845.

**Comment by FelixGruetzmacher** [comment]
Don't have a working copy here as I'm no longer at the office but I looked at the code and I think you got it nailed. "hwPortUtils.pyc", line 236 tries to log a Windows error which might, depending on locale, contain unexpected characters. Maybe something as simple as using %r instead of %s gets rid of it.

**Comment by leonardder** [comment]
Could you report whether [this](https://ci.appveyor.com/api/buildjobs/c77qm6overooi51d/artifacts/output%2Fnvda_snapshot_try-handytech-native-14471%2C7dd31e6b.exe) try build fixes your issue? it uses %r instead of %s to log the port.

**Comment by FelixGruetzmacher** [comment]
Positive!

Leonard de Ruijter <notifications@github.com> schrieb am Fr., 29. Sep. 2017
um 21:46 Uhr:

**Comment by FelixGruetzmacher** [comment]
Any ideas if this will make it to 2017.4?

**Comment by bramd** [comment]
[**@FelixGruetzmacher**](https://github.com/FelixGruetzmacher) commented on [Oct 1, 2017, 10:30 AM GMT+2](https://github.com/nvaccess/nvda/issues/7458#issuecomment-333361655 "2017-10-01T08:30:04Z - Replied by Github Reply Comments"):
> Any ideas if this will make it to 2017.4?

It is not yet reviewed by one of the core devs and will need to go in next for testing first, before it will move on to master and a release. Depending on that timeline and if we find any bugs when it's in next, this might or might not be in 2017.4.

Also, I think it would be good to have a clear path for users willing or needing to use the COM server. Bundling both would be confusing for users, so I think it would be good to provide the COM server as an add-on. Is the Handy Tech add-on that already exists available on the website? I only saw it on the FTP site and not in the downloads section. Also, we might investigate of that add-on can register the COM server on installation if it hasn't been registered already, so users don't have to run the setup for the universal COM driver.

**Comment by derekriemer** [comment]
Could it be registered the first time the user launches the display?

**Comment by leonardder** [comment]
I belief the com server dll can now be bundled since the com server is specified in NVDA's manifest. I'm afraid it won't load if we leave it out there.

**Comment by FelixGruetzmacher** [comment]
Maybe if there's a way to modify the activation context at runtime without
having the com server in the manifest. The alternative might be to create a
setup which registers the com server with Windows and installs the add-on.
It can be written so as not to interfere with the native driver by simply
naming it differently, such as "Handy Tech Custom" as opposed to "Handy
Tech".

Leonard de Ruijter <notifications@github.com> schrieb am Mo., 2. Okt. 2017
um 06:15 Uhr:

**Comment by leonardder** [comment]
[**@FelixGruetzmacher**](https://github.com/FelixGruetzmacher) commented on [2 okt. 2017 11:42 CEST](https://github.com/nvaccess/nvda/issues/7458#issuecomment-333487512 "2017-10-02T09:42:23Z - Replied by Github Reply Comments"):
> The alternative might be to create a
> setup which registers the com server with Windows and installs the add-on.
> It can be written so as not to interfere with the native driver by simply
> naming it differently, such as "Handy Tech Custom" as opposed to "Handy
> Tech".

That would be something you guys at handy tech could do of course. IN any case, I think everything related to the com server should be removed from NVDA core.

**Comment by FelixGruetzmacher** [comment]
I agree. And regarding said setup, I meant it in exactly the way you
understood it---namely, that we at Handy Tech would do that. I just wanted
to point out we would make very sure it plays well with the native
implementation, which I like a lot.

Leonard de Ruijter <notifications@github.com> schrieb am Mo., 2. Okt. 2017
um 12:01 Uhr:
